core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/LimitedListLineConsumer.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.core.util;
+
+/**
+ * Extension of {@link ListLineConsumer} which consumes limited amount of text {@link #writeLine}.
+ *
+ * <p>Implementation is not threadsafe and requires external synchronization if is used in
+ * multi-thread environment.
+ *
+ * @author Dmytro Nochevnov
+ */
+public class LimitedListLineConsumer extends ListLineConsumer {
+  private final long textSizeLimit;
+  private long consumedTextSize;
+
+  /**
+   * Limits size of consuming text.
+   *
+   * @param textSizeLimit maximum number of symbols of consuming text.
+   * Don't limit text if {@code textSizeLimit} < 0.
+   * Don't consume if {@code textSizeLimit} = 0.
+   */
+  public LimitedListLineConsumer(long textSizeLimit) {
+    super();
+    this.textSizeLimit = textSizeLimit;
+  }
+
+  @Override
+  public void writeLine(String line) {
+    if (textSizeLimit < 0) {
+      lines.add(line);
+      return;
+    }
+
+    if (consumedTextSize >= textSizeLimit) {
```

Comment:
Very strange algorithm. Can you describe in general how writeLine of this class is working?

Reply:
- _writeLine()_ ensures that size of consuming text, that `([line1 size] + [line2 size] + ... + [lineN size]) <= [textSizeLimit value]` if **textSizeLimit** > 0
- _writeLine()_ doesn't limit size if **textSizeLimit** < 0
- _writeLine()_ ignores lines if **textSizeLimit** = 0
- _writeLine()_ cut line if `consumedTextSize + [lineN size]) > [textSizeLimit value]` so that `consumedTextSize + [cut lineN size]) = [textSizeLimit value]`

Reply:
Personally, It should look like this:
```
public class LimitedSizeLineConsumer implements LineConsumer {
  private final LineConsumer lineConsumer;
  private final int capacity;
  private int length;

  public LimitedSizeLineConsumer(LineConsumer lineConsumer, int capacity) {
    this.lineConsumer = lineConsumer;
    this.capacity = capacity;
  }

  @Override
  public void writeLine(String line) throws IOException {
    if (line == null) {
      return;
    }

    int maxLineLength = Math.min(capacity - length, line.length());
    lineConsumer.writeLine(line.substring(0, maxLineLength));
    length += maxLineLength;
  }

  @Override
  public void close() throws IOException {
    lineConsumer.close();
  }
}
```

Take into account that, `substring` method doesn't create a new string, so the original string still will be in memory.

Reply:
I agree with @tolusha. His implementation looks better. @dmytro-ndp your implementation looks overcomplicated. The thing that I would add is a condition if we know that we already wrote the maximum number of characters then we simply exit from ```writeLine``` method.

@dmytro-ndp Please consider moving LimitedSizeLineConsumer to selenium framework.

Reply:
@tolusha: your implementation looks simpler, and I like an idea to have a kind of wrapper for **LineConsumer** classes. Despite that I see some issues there:
1. It doesn't check that capacity is positive number. But it could be easily fixed.
1. It doesn't work properly in case of usage **org.eclipse.che.api.core.util.ListLineConsumer** after **org.eclipse.che.api.core.util.ListLineConsumer#clear()** method is called.

Reply:
@skabashnyuk: 
> The thing that I would add is a condition if we know that we already wrote the maximum number of characters then we simply exit from `writeLine` method.
Good catch. Thank you!